### PR TITLE
[config] Fix 'config reload -l' command to get filename by default

### PR DESCRIPTION
**Pull request by samaity**
Signed-off-by: Sangita Maity <samaity@linkedin.com>

#### What I did

Fix https://github.com/Azure/sonic-buildimage/issues/7433

Right now `config reload -l` is getting failed due to below error.
```
admin@lnos-x1-a-asw03:~$ sudo config reload -l
Clear current config and reload config from the default config file(s) ? [y/N]: y
Traceback (most recent call last):
  File "/usr/local/bin/sonic-cfggen", line 431, in <module>
    main()
  File "/usr/local/bin/sonic-cfggen", line 326, in main
    _process_json(args, data)
  File "/usr/local/bin/sonic-cfggen", line 236, in _process_json
    with open(json_file, 'r') as stream:
FileNotFoundError: [Errno 2] No such file or directory: 'None'
Disabling container monitoring ...
Stopping SONiC target ...
Running command: /usr/local/bin/sonic-cfggen -H -k  --write-to-db
usage: sonic-cfggen [-h] [-m [MINIGRAPH] | -M DEVICE_DESCRIPTION | -k HWSKU] [-n [NAMESPACE]] [-p [PORT_CONFIG]]
                    [-S [HWSKU_CONFIG]] [-y YAML] [-j JSON] [-a ADDITIONAL_DATA] [-d] [-H] [-s REDIS_UNIX_SOCK_FILE]
                    [-t TEMPLATE] [-T TEMPLATE_DIR] [-v VAR] [--var-json VAR_JSON] [--preset {t1,l2,empty}]
                    [--print-data | -w | -K KEY]
sonic-cfggen: error: argument -k/--hwsku: expected one argument

```

I guess the problem is [here](https://github.com/Azure/sonic-utilities/blob/master/config/main.py#L1189) in `sonic-utilities` repo. If user does not provide filename with `config reload -l`,` command = "{} -j {} -v DEVICE_METADATA.localhost.hwsku".format(SONIC_CFGGEN_PATH, filename)` will not provide `cfg_hwsku ` i.e. hwsku parameter as it should which will later cause problem [here](https://github.com/Azure/sonic-utilities/blob/master/config/main.py#L1245) `command = "{} -H -k {} --write-to-db".format(SONIC_CFGGEN_PATH, cfg_hwsku)` as hwsku is not available around that time.

that's why we notice errors like `No such file or directory: 'None'` as pasted in this issue.

```
File "/usr/local/bin/sonic-cfggen", line 328, in main
    _process_json(args, data)
  File "/usr/local/bin/sonic-cfggen", line 236, in _process_json
    with open(json_file, 'r') as stream:
FileNotFoundError: [Errno 2] No such file or directory: 'None'
```

#### How I did it

To Fix the issue, moved  the part where the code gets cfg_hwsku `command = "{} -j {} -v DEVICE_METADATA.localhost.hwsku".format(SONIC_CFGGEN_PATH, file)` to the same location it needed as we get filename by default.

#### How to verify it

**Before the Fix**
```
admin@lnos-x1-a-asw03:~$ sudo config reload -l
Clear current config and reload config from the default config file(s) ? [y/N]: y
Traceback (most recent call last):
  File "/usr/local/bin/sonic-cfggen", line 431, in <module>
    main()
  File "/usr/local/bin/sonic-cfggen", line 326, in main
    _process_json(args, data)
  File "/usr/local/bin/sonic-cfggen", line 236, in _process_json
    with open(json_file, 'r') as stream:
FileNotFoundError: [Errno 2] No such file or directory: 'None'
Disabling container monitoring ...
Stopping SONiC target ...
```
**After the fix**
```
admin@lnos-x1-a-asw03:~$ sudo config reload -l
Clear current config and reload config from the default config file(s) ? [y/N]: y
Disabling container monitoring ...
Stopping SONiC target ...
Running command: /usr/local/bin/sonic-cfggen -H -k Seastone-DX010-25-50 --write-to-db
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
Running command: /usr/local/bin/db_migrator.py -o migrate
Resetting failed status on bgp.service
Resetting failed status on caclmgrd.service
Resetting failed status on dhcp_relay.service
Resetting failed status on hostcfgd.service
Resetting failed status on hostname-config.service
Resetting failed status on interfaces-config.service
Resetting failed status on lldp.service
Resetting failed status on mgmt-framework.timer
Resetting failed status on nat.service
Resetting failed status on ntp-config.service
Resetting failed status on pmon.service
Resetting failed status on procdockerstatsd.service
Resetting failed status on radv.service
Resetting failed status on rsyslog-config.service
Resetting failed status on sflow.service
Resetting failed status on swss.service
Resetting failed status on syncd.service
Resetting failed status on teamd.service
Resetting failed status on telemetry.timer
Restarting SONiC target ...
Enabling container monitoring ...
Reloading Monit configuration ...
Reinitializing monit daemon
```

Added Test cases.
```
samaity@973da139a792:/sonic/src/sonic-utilities$ python3 setup.py test --addopts "--capture=no tests/config_test.py::TestConfigReload::test_config_reload"
running pytest
 
running build_ext
 
================================================================================================== test session starts ===================================================================================================
platform linux -- Python 3.7.3, pytest-3.10.1, py-1.7.0, pluggy-0.8.0
rootdir: /sonic/src/sonic-utilities/tests, inifile: pytest.ini
plugins: pyfakefs-4.5.0, cov-2.6.0
collected 1 item
 
tests/config_test.py SETUP
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
0
Running command: rm -rf /tmp/dropstat-*
Running command: /usr/local/bin/sonic-cfggen -H -k Seastone-DX010-25-50 --write-to-db
Running command: /usr/local/bin/sonic-cfggen -j /sonic/src/sonic-utilities/tests/config_reload_input/init_cfg.json -j /sonic/src/sonic-utilities/tests/config_reload_input/config_db.json --write-to-db
 
  File "/sonic/src/sonic-utilities/.eggs/Click-7.0-py3.7.egg/click/testing.py", line 326, in invoke
    cli.main(args=args or (), prog_name=prog_name, **extra)
  File "/sonic/src/sonic-utilities/.eggs/Click-7.0-py3.7.egg/click/core.py", line 745, in main
    sys.exit(e.exit_code)
.TEARDOWN
 
================================================================= 1 passed in 1.19 seconds =================================================================
samaity@973da139a792:/sonic/src/sonic-utilities$

```

### CODE REVIEW

**Review by zhenggen-xu — DISMISSED** [review]

**Review by liat-grozovik — DISMISSED** [review]

File: config/main.py [hunk 1]
```diff
@@ -1179,7 +1169,17 @@ def reload(db, filename, yes, load_sysinfo, no_service_restart, disable_arp_cach
...
+
         if load_sysinfo:
+            command = "{} -j {} -v DEVICE_METADATA.localhost.hwsku".format(SONIC_CFGGEN_PATH, file)
+            proc = subprocess.Popen(command, shell=True, text=True, stdout=subprocess.PIPE)
+            cfg_hwsku, err = proc.communicate()
+            if err:
+                click.echo("Could not get the HWSKU from config file, exiting")
+                sys.exit(1)
+            else:
+                cfg_hwsku = cfg_hwsku.decode('UTF-8').strip()
```
Comment (qiluo-msft) [inline config/main.py:1181]: >'UTF-8' [](http://example.com/codeflow?start=45&length=7)

Not a blocking issue, but you can remove it and use the default value.

**Review by qiluo-msft — DISMISSED** [review]

**Comment by lguohan** [comment]
please add unit test to prevent futher regression. which pr caused the regression?

**Comment by zhenggen-xu** [comment]
Based on the log, it should be this PR that caused regression: https://github.com/Azure/sonic-utilities/pull/877

File: config/main.py (hunk 1 above)
Comment (samaity) [reply config/main.py]: Modified as requested.

**Comment by samaity** [comment]
> please add unit test to prevent further regression. which pr caused the regression?

Added Test cases as requested.

File: config/main.py [hunk 2]
```diff
@@ -1322,6 +1312,26 @@ def reload(db, filename, yes, load_sysinfo, no_service_restart, disable_arp_cach
             click.echo("The config_db file {} doesn't exist".format(file))
             continue
 
+        if load_sysinfo:
+            try:
+                command = "{} -j {} -v DEVICE_METADATA.localhost.hwsku".format(SONIC_CFGGEN_PATH, file)
+                #output = clicommon.run_command(command, display_cmd= True, return_cmd=True)
```
Comment (qiluo-msft) [inline config/main.py:1318]: >#output = clicommon.run_command(command, display_cmd= True, return_cmd=True) [](http://example.com/codeflow?start=16&length=76)

Remove dead code

**Review by zhenggen-xu — APPROVED** [review]

File: config/main.py (hunk 2 above)
Comment (samaity) [reply config/main.py]: Modified as requested

**Review by qiluo-msft — APPROVED** [review]

**Comment by dgsudharsan** [comment]
@samaity Can you please share the status of this PR? When will this be merged?

**Comment by zhenggen-xu** [comment]
@samaity  Can you sync to the latest code?

**Comment by samaity** [comment]
> @samaity Can you sync to the latest code?

done 👍

**Review by zhenggen-xu — APPROVED** [review]

**Pull request merged**